File: optimizer.py
"""Distributor-agnostic purchase-plan optimization helpers.

Supplier adapters normalize their packaging and price-break information into
``PurchaseFamily`` records. This module then handles the actual buy-plan
optimization, including exact buys, overbuy via larger price breaks, and
mixed bulk-plus-remainder plans such as ``3 reels + 600 cut tape``.
"""

from __future__ import annotations

from dataclasses import dataclass
from math import ceil
from typing import Iterable

from models import PurchaseLeg


@dataclass(frozen=True)
class FamilyPriceBreak:
    """One normalized price break inside a purchase family."""

    quantity: int
    unit_price: float
    currency: str


@dataclass(frozen=True)
class PurchaseFamily:
    """One normalized purchasable packaging family from a distributor."""

    family_id: str
    package_type: str | None = None
    packaging_mode: str | None = None
    minimum_order_quantity: int | None = None
    order_multiple: int | None = None
    full_reel_quantity: int | None = None
    base_pricing_strategy: str | None = None
    strategy_mode: str = "static"
    allow_mixing_as_bulk: bool = False
    allow_mixing_as_remainder: bool = True
    mix_quantity: int | None = None
    price_breaks: tuple[FamilyPriceBreak, ...] = ()


@dataclass(frozen=True)
class OptimizedPurchasePlan:
    """A fully-priced purchase plan, possibly composed from multiple legs."""

    required_quantity: int
    purchased_quantity: int
    unit_price: float
    extended_price: float
    currency: str
    price_break_quantity: int | None
    surplus_quantity: int
    pricing_strategy: str
    order_plan: str | None
    purchase_legs: tuple[PurchaseLeg, ...]


def _round_up_to_multiple(quantity: int, multiple: int | None) -> int:
    """Round a quantity up to the next legal multiple when needed."""
    if multiple is None or multiple <= 1:
        return quantity
    return int(ceil(quantity / multiple) * multiple)


def _batch_noun(leg: PurchaseLeg) -> str:
    """Return the preferred noun for one purchase leg's packaging."""
    packaging_text = _packaging_text(leg).lower()
    if "reel" in packaging_text and "cut tape" not in packaging_text and "mousereel" not in packaging_text:
        return "reel"
    if "tray" in packaging_text:
        return "tray"
    if "tube" in packaging_text:
        return "tube"
    return "batch"


def _packaging_text(leg: PurchaseLeg) -> str:
    """Return one de-duplicated packaging label string for a purchase leg."""
    texts: list[str] = []
    seen: set[str] = set()
    for text in [leg.packaging_mode, leg.package_type]:
        normalized = (text or "").strip()
        if not normalized:
            continue
        lowered = normalized.lower()
        if lowered in seen:
            continue
        texts.append(normalized)
        seen.add(lowered)
    return " ".join(texts)


def format_purchase_leg(leg: PurchaseLeg) -> str:
    """Return a short human-readable string for one concrete purchase leg."""
    if leg.order_batch_quantity and leg.order_batch_count:
        noun = _batch_noun(leg)
        plural = noun if leg.order_batch_count == 1 else f"{noun}s"
        return f"{leg.order_batch_count} {plural} x {leg.order_batch_quantity}"

    packaging_text = _packaging_text(leg)
    if packaging_text:
        return f"{leg.purchased_quantity} {packaging_text.lower()}"
    return f"{leg.purchased_quantity}"


def format_order_plan(legs: Iterable[PurchaseLeg]) -> str:
    """Return the combined display string for one or more purchase legs."""
    rendered = [format_purchase_leg(leg) for leg in legs]
    return " + ".join(item for item in rendered if item)


def compose_purchase_plan(
    required_quantity: int,
    legs: Iterable[PurchaseLeg],
    pricing_strategy: str,
    *,
    price_break_quantity: int | None = None,
    order_plan: str | None = None,
) -> OptimizedPurchasePlan | None:
    """Aggregate one or more priced legs into a comparable purchase plan."""
    purchase_legs = tuple(legs)
    if not purchase_legs:
        return None

    currencies = {leg.currency for leg in purchase_legs if leg.currency}
    if len(currencies) != 1:
        return None

    purchased_quantity = sum(leg.purchased_quantity for leg in purchase_legs)
    extended_price = round(sum(leg.extended_price for leg in purchase_legs), 2)
    unit_price = round(extended_price / purchased_quantity, 6) if purchased_quantity else 0.0
    price_break = price_break_quantity
    if price_break is None and len(purchase_legs) == 1:
        price_break = purchase_legs[0].price_break_quantity

    return OptimizedPurchasePlan(
        required_quantity=required_quantity,
        purchased_quantity=purchased_quantity,
        unit_price=unit_price,
        extended_price=extended_price,
        currency=next(iter(currencies), ""),
        price_break_quantity=price_break,
        surplus_quantity=max(purchased_quantity - required_quantity, 0),
        pricing_strategy=pricing_strategy,
        order_plan=order_plan or format_order_plan(purchase_legs),
        purchase_legs=purchase_legs,
    )


def select_best_purchase_plan(
    plans: Iterable[OptimizedPurchasePlan],
) -> OptimizedPurchasePlan | None:
    """Return the cheapest valid plan using stable tiebreakers."""
    plan_list = list(plans)
    if not plan_list:
        return None
    return min(
        plan_list,
        key=lambda plan: (
            plan.extended_price,
            plan.surplus_quantity,
            plan.purchased_quantity,
            len(plan.purchase_legs),
            float("inf") if plan.price_break_quantity is None else plan.price_break_quantity,
        ),
    )


def _family_strategy(
    family: PurchaseFamily,
    *,
    required_quantity: int,
    purchased_quantity: int,
    break_quantity: int,
) -> str:
    """Return the strategy label for one family-specific buy leg."""
    if family.strategy_mode == "static":
        return family.base_pricing_strategy or ""
    if family.strategy_mode == "full_reel":
        return "full reel"
    if family.strategy_mode == "price_break":
        if purchased_quantity > required_quantity:
            if family.order_multiple and purchased_quantity % family.order_multiple == 0:
                return "order multiple" if break_quantity <= required_quantity else "next price break"
            return "next price break"
        return family.base_pricing_strategy or "requested quantity"
    return family.base_pricing_strategy or ""


def purchase_leg_from_family(
    family: PurchaseFamily,
    quantity: int,
) -> PurchaseLeg | None:
    """Return the cheapest legal single-family leg for one quantity target."""
    if quantity <= 0:
        return None

    legs: list[PurchaseLeg] = []
    for price_break in family.price_breaks:
        break_quantity = int(price_break.quantity)
        if break_quantity <= 0:
            continue

        base_quantity = max(quantity, break_quantity, family.minimum_order_quantity or 0)
        rounding_multiple = family.full_reel_quantity or family.order_multiple
        purchased_quantity = _round_up_to_multiple(base_quantity, rounding_multiple)
        purchased_quantity = max(purchased_quantity, base_quantity)

        order_batch_quantity = next(
            (
                candidate
                for candidate in [
                    family.full_reel_quantity,
                    family.order_multiple,
                    family.minimum_order_quantity,
                ]
                if candidate is not None and candidate > 1 and purchased_quantity % candidate == 0
            ),
            None,
        )
        order_batch_count = (
            purchased_quantity // order_batch_quantity
            if order_batch_quantity is not None
            else None
        )
        legs.append(
            PurchaseLeg(
                purchased_quantity=purchased_quantity,
                unit_price=price_break.unit_price,
                extended_price=round(price_break.unit_price * purchased_quantity, 2),
                currency=price_break.currency,
                price_break_quantity=break_quantity,
                pricing_strategy=_family_strategy(
                    family,
                    required_quantity=quantity,
                    purchased_quantity=purchased_quantity,
                    break_quantity=break_quantity,
                ),
                package_type=family.package_type,
                packaging_mode=family.packaging_mode,
                order_batch_quantity=order_batch_quantity,
                order_batch_count=order_batch_count,
            )
        )

    if not legs:
        return None

    return min(
        legs,
        key=lambda leg: (
            leg.extended_price,
            leg.purchased_quantity,
            float("inf") if leg.price_break_quantity is None else leg.price_break_quantity,
        ),
    )


def purchase_plan_from_family(
    family: PurchaseFamily,
    quantity: int,
) -> OptimizedPurchasePlan | None:
    """Return the cheapest legal single-family plan for one quantity target."""
    leg = purchase_leg_from_family(family, quantity)
    if leg is None:
        return None
    return compose_purchase_plan(
        quantity,
        [leg],
        leg.pricing_strategy or family.base_pricing_strategy or "",
    )


def _candidate_bulk_quantities(
    required_quantity: int,
    family: PurchaseFamily,
) -> tuple[int, ...]:
    """Return the candidate bulk quantities worth testing for mixed plans."""
    mix_quantity = family.mix_quantity or family.full_reel_quantity or family.order_multiple
    if mix_quantity is None or mix_quantity <= 1:
        return ()

    max_break_quantity = max((price_break.quantity for price_break in family.price_breaks), default=0)
    max_quantity = max(required_quantity, max_break_quantity)
    max_count = max(1, ceil(max_quantity / mix_quantity))
    return tuple(count * mix_quantity for count in range(1, max_count + 1))


def optimize_purchase_families(
    required_quantity: int,
    families: Iterable[PurchaseFamily],
    *,
    mixed_strategy: str = "mixed packaging",
) -> OptimizedPurchasePlan | None:
    """Return the best plan across all single-family and mixed-family options."""
    family_list = [family for family in families if family.price_breaks]
    plans: list[OptimizedPurchasePlan] = []

    for family in family_list:
        plan = purchase_plan_from_family(family, required_quantity)
        if plan is not None:
            plans.append(plan)

    bulk_families = [family for family in family_list if family.allow_mixing_as_bulk]
    remainder_families = [family for family in family_list if family.allow_mixing_as_remainder]

    for bulk_family in bulk_families:
        for bulk_quantity in _candidate_bulk_quantities(required_quantity, bulk_family):
            bulk_leg = purchase_leg_from_family(bulk_family, bulk_quantity)
            if bulk_leg is None:
                continue

            remainder_quantity = max(required_quantity - bulk_leg.purchased_quantity, 0)
            if remainder_quantity <= 0:
                continue

            for remainder_family in remainder_families:
                if remainder_family.family_id == bulk_family.family_id:
                    continue
                remainder_leg = purchase_leg_from_family(remainder_family, remainder_quantity)
                if remainder_leg is None:
                    continue
                plan = compose_purchase_plan(
                    required_quantity,
                    [bulk_leg, remainder_leg],
                    mixed_strategy,
                )
                if plan is not None:
                    plans.append(plan)

    return select_best_purchase_plan(plans)

File: internal/procurement/optimizer.go
// Copyright (C) 2026 Joern Ihlenburg
// SPDX-License-Identifier: GPL-3.0-or-later

// Package procurement defines normalized distributor offers and safe purchasing plans.
package procurement

import (
	"errors"
	"fmt"
	"math"
	"math/big"
	"sort"
	"strings"

	"github.com/jihlenburg/bom-builder/internal/money"
)

const defaultManufacturingPreferenceBasisPoints = 50 // 0.50%

// PriceBreak is one exact unit-price threshold.
type PriceBreak struct {
	Quantity  int           `json:"quantity"`
	UnitPrice money.Decimal `json:"unit_price"`
	Currency  string        `json:"currency"`
}

// PurchaseFamily describes one independently orderable packaging family.
type PurchaseFamily struct {
	ID                     string
	PackageType            string
	PackagingMode          string
	MinimumOrderQuantity   int
	OrderMultiple          int
	FullReelQuantity       int
	BasePricingStrategy    string
	StrategyMode           string
	AllowMixingAsBulk      bool
	AllowMixingAsRemainder bool
	MixQuantity            int
	AvailableQuantity      *int
	PriceBreaks            []PriceBreak
}

// PurchaseLeg is one distributor SKU/packaging leg in an order plan.
type PurchaseLeg struct {
	FamilyID             string        `json:"family_id"`
	PurchasedQuantity    int           `json:"purchased_quantity"`
	UnitPrice            money.Decimal `json:"unit_price"`
	ExtendedPrice        money.Decimal `json:"extended_price"`
	Currency             string        `json:"currency"`
	PriceBreak           int           `json:"price_break_quantity"`
	PricingStrategy      string        `json:"pricing_strategy"`
	PackageType          string        `json:"package_type,omitempty"`
	PackagingMode        string        `json:"packaging_mode,omitempty"`
	MinimumOrderQuantity int           `json:"minimum_order_quantity,omitempty"`
	OrderMultiple        int           `json:"order_multiple,omitempty"`
	OrderBatchSize       int           `json:"order_batch_quantity,omitempty"`
	OrderBatchCount      int           `json:"order_batch_count,omitempty"`
	Marketplace          bool          `json:"marketplace"`
	StockVerified        bool          `json:"stock_verified"`
}

// PurchasePlan is a stock-aware, exactly priced plan covering a requirement.
type PurchasePlan struct {
	RequiredQuantity  int           `json:"required_quantity"`
	PurchasedQuantity int           `json:"purchased_quantity"`
	SurplusQuantity   int           `json:"surplus_quantity"`
	UnitPrice         money.Decimal `json:"effective_unit_price"`
	ExtendedPrice     money.Decimal `json:"extended_price"`
	Currency          string        `json:"currency"`
	PriceBreak        int           `json:"price_break_quantity,omitempty"`
	PricingStrategy   string        `json:"pricing_strategy"`
	OrderPlan         string        `json:"order_plan"`
	StockVerified     bool          `json:"stock_verified"`
	Legs              []PurchaseLeg `json:"legs"`
}

// OptimizePurchaseFamilies returns the preferred legal single- or mixed-family plan.
func OptimizePurchaseFamilies(requiredQuantity int, families []PurchaseFamily) (*PurchasePlan, error) {
	return OptimizePurchaseFamiliesWithPreference(
		requiredQuantity,
		families,
		defaultManufacturingPreferenceBasisPoints,
	)
}

// OptimizePurchaseFamiliesWithPreference allows manufacturing preference to be tested explicitly.
func OptimizePurchaseFamiliesWithPreference(
	requiredQuantity int,
	families []PurchaseFamily,
	preferenceBasisPoints int,
) (*PurchasePlan, error) {
	if requiredQuantity < 1 {
		return nil, errors.New("required quantity must be positive")
	}
	if preferenceBasisPoints < 0 {
		return nil, errors.New("manufacturing preference cannot be negative")
	}

	validFamilies := make([]PurchaseFamily, 0, len(families))
	for _, family := range families {
		if len(family.PriceBreaks) == 0 {
			continue
		}
		if err := validateFamily(family); err != nil {
			return nil, fmt.Errorf("purchase family %q: %w", family.ID, err)
		}
		validFamilies = append(validFamilies, family)
	}

	// Every candidate plan is later compared by raw ExtendedPrice micros,
	// which is meaningless across currencies: a silent pick could select
	// the economically more expensive plan. validateFamily guarantees one
	// currency per family; this guarantees one currency per comparison,
	// mirroring the sourcing layer's explicit conversion refusal.
	currencies := make(map[string]bool)
	for _, family := range validFamilies {
		currencies[family.PriceBreaks[0].Currency] = true
	}
	if len(currencies) > 1 {
		names := make([]string, 0, len(currencies))
		for currency := range currencies {
			names = append(names, currency)
		}
		sort.Strings(names)
		return nil, fmt.Errorf(
			"purchase families span multiple currencies (%s): cross-currency comparison requires explicit conversion",
			strings.Join(names, ", "),
		)
	}

	plans := make([]PurchasePlan, 0, len(validFamilies))
	for _, family := range validFamilies {
		leg, err := purchaseLegFromFamily(family, requiredQuantity)
		if err != nil {
			return nil, err
		}
		if leg != nil {
			plan, err := composePurchasePlan(
				requiredQuantity,
				[]PurchaseLeg{*leg},
				leg.PricingStrategy,
			)
			if err != nil {
				return nil, err
			}
			plans = append(plans, *plan)
		}
	}

	for _, bulk := range validFamilies {
		if !bulk.AllowMixingAsBulk {
			continue
		}
		for _, remainder := range validFamilies {
			if !remainder.AllowMixingAsRemainder || remainder.ID == bulk.ID {
				continue
			}
			for _, quantity := range candidateBulkQuantities(requiredQuantity, bulk, remainder) {
				bulkLeg, err := purchaseLegFromFamily(bulk, quantity)
				if err != nil {
					return nil, err
				}
				if bulkLeg == nil || bulkLeg.PurchasedQuantity >= requiredQuantity {
					continue
				}
				remainderLeg, err := purchaseLegFromFamily(
					remainder,
					requiredQuantity-bulkLeg.PurchasedQuantity,
				)
				if err != nil {
					return nil, err
				}
				if remainderLeg == nil {
					continue
				}
				plan, err := composePurchasePlan(
					requiredQuantity,
					[]PurchaseLeg{*bulkLeg, *remainderLeg},
					"mixed packaging",
				)
				if err != nil {
					return nil, err
				}
				plans = append(plans, *plan)
			}
		}
	}

	if len(plans) == 0 {
		return nil, nil
	}
	return selectBestPlan(plans, preferenceBasisPoints), nil
}

func validateFamily(family PurchaseFamily) error {
	if strings.TrimSpace(family.ID) == "" {
		return errors.New("ID is required")
	}
	currency := ""
	for _, value := range []struct {
		name  string
		value int
	}{
		{"minimum order quantity", family.MinimumOrderQuantity},
		{"order multiple", family.OrderMultiple},
		{"full reel quantity", family.FullReelQuantity},
		{"mix quantity", family.MixQuantity},
	} {
		if value.value < 0 {
			return fmt.Errorf("%s cannot be negative", value.name)
		}
	}
	if family.AvailableQuantity != nil && *family.AvailableQuantity < 0 {
		return errors.New("available quantity cannot be negative")
	}
	for _, priceBreak := range family.PriceBreaks {
		if priceBreak.Quantity < 1 {
			return errors.New("price-break quantity must be positive")
		}
		if priceBreak.UnitPrice.Micros() <= 0 {
			return errors.New("unit price must be positive")
		}
		if !validCurrency(priceBreak.Currency) {
			return fmt.Errorf("invalid currency %q", priceBreak.Currency)
		}
		if currency == "" {
			currency = priceBreak.Currency
		} else if priceBreak.Currency != currency {
			return errors.New("price breaks contain mixed currencies")
		}
	}
	return nil
}

func purchaseLegFromFamily(family PurchaseFamily, quantity int) (*PurchaseLeg, error) {
	if quantity <= 0 {
		return nil, nil
	}
	var candidates []PurchaseLeg
	for _, priceBreak := range family.PriceBreaks {
		purchased := max(quantity, priceBreak.Quantity, family.MinimumOrderQuantity)
		multiple := family.OrderMultiple
		if family.FullReelQuantity > 0 {
			multiple = family.FullReelQuantity
		}
		var err error
		purchased, err = roundUpToMultiple(purchased, multiple)
		if err != nil {
			return nil, err
		}
		if family.AvailableQuantity != nil && purchased > *family.AvailableQuantity {
			continue
		}
		extended, err := priceBreak.UnitPrice.MulInt(purchased)
		if err != nil {
			return nil, err
		}
		batchSize := firstDivisor(
			purchased,
			family.FullReelQuantity,
			family.OrderMultiple,
			family.MinimumOrderQuantity,
		)
		batchCount := 0
		if batchSize > 1 {
			batchCount = purchased / batchSize
		}
		candidates = append(candidates, PurchaseLeg{
			FamilyID:          family.ID,
			PurchasedQuantity: purchased,
			UnitPrice:         priceBreak.UnitPrice,
			ExtendedPrice:     extended,
			Currency:          strings.ToUpper(priceBreak.Currency),
			PriceBreak:        priceBreak.Quantity,
			PricingStrategy: familyStrategy(
				family,
				quantity,
				purchased,
				priceBreak.Quantity,
			),
			PackageType:          family.PackageType,
			PackagingMode:        family.PackagingMode,
			MinimumOrderQuantity: family.MinimumOrderQuantity,
			OrderMultiple:        family.OrderMultiple,
			OrderBatchSize:       batchSize,
			OrderBatchCount:      batchCount,
			StockVerified:        family.AvailableQuantity != nil,
		})
	}
	if len(candidates) == 0 {
		return nil, nil
	}
	sort.SliceStable(candidates, func(left, right int) bool {
		return compareLeg(candidates[left], candidates[right]) < 0
	})
	return &candidates[0], nil
}

func composePurchasePlan(
	requiredQuantity int,
	legs []PurchaseLeg,
	strategy string,
) (*PurchasePlan, error) {
	if len(legs) == 0 {
		return nil, nil
	}
	currency := legs[0].Currency
	purchased := 0
	total := money.Decimal(0)
	stockVerified := true
	for _, leg := range legs {
		if leg.Currency != currency {
			return nil, errors.New("purchase plan contains mixed currencies")
		}
		var err error
		total, err = total.Add(leg.ExtendedPrice)
		if err != nil {
			return nil, err
		}
		if leg.PurchasedQuantity > math.MaxInt-purchased {
			return nil, errors.New("purchase quantity overflow")
		}
		purchased += leg.PurchasedQuantity
		stockVerified = stockVerified && leg.StockVerified
	}
	if purchased < requiredQuantity {
		return nil, errors.New("purchase plan does not cover required quantity")
	}
	effective, err := total.DivInt(purchased)
	if err != nil {
		return nil, err
	}
	priceBreak := 0
	if len(legs) == 1 {
		priceBreak = legs[0].PriceBreak
	}
	return &PurchasePlan{
		RequiredQuantity:  requiredQuantity,
		PurchasedQuantity: purchased,
		SurplusQuantity:   purchased - requiredQuantity,
		UnitPrice:         effective,
		ExtendedPrice:     total,
		Currency:          currency,
		PriceBreak:        priceBreak,
		PricingStrategy:   strategy,
		OrderPlan:         formatOrderPlan(legs),
		StockVerified:     stockVerified,
		Legs:              legs,
	}, nil
}

func selectBestPlan(plans []PurchasePlan, preferenceBasisPoints int) *PurchasePlan {
	sort.SliceStable(plans, func(left, right int) bool {
		return comparePlanCost(plans[left], plans[right]) < 0
	})
	cheapest := plans[0]
	preferred := make([]PurchasePlan, 0, len(plans))
	for _, plan := range plans {
		if withinBasisPoints(plan.ExtendedPrice, cheapest.ExtendedPrice, preferenceBasisPoints) {
			preferred = append(preferred, plan)
		}
	}
	sort.SliceStable(preferred, func(left, right int) bool {
		return compareManufacturingPreference(preferred[left], preferred[right]) < 0
	})
	selected := preferred[0]
	return &selected
}

func compareLeg(left, right PurchaseLeg) int {
	if result := cmpDecimal(left.ExtendedPrice, right.ExtendedPrice); result != 0 {
		return result
	}
	if left.PurchasedQuantity != right.PurchasedQuantity {
		return cmpInt(left.PurchasedQuantity, right.PurchasedQuantity)
	}
	return cmpInt(left.PriceBreak, right.PriceBreak)
}

func comparePlanCost(left, right PurchasePlan) int {
	if result := cmpDecimal(left.ExtendedPrice, right.ExtendedPrice); result != 0 {
		return result
	}
	if left.SurplusQuantity != right.SurplusQuantity {
		return cmpInt(left.SurplusQuantity, right.SurplusQuantity)
	}
	if left.PurchasedQuantity != right.PurchasedQuantity {
		return cmpInt(left.PurchasedQuantity, right.PurchasedQuantity)
	}
	return cmpInt(len(left.Legs), len(right.Legs))
}

func compareManufacturingPreference(left, right PurchasePlan) int {
	leftReel, leftStable, leftCut := packagingQuantities(left)
	rightReel, rightStable, rightCut := packagingQuantities(right)
	for _, comparison := range []int{
		cmpInt(rightReel, leftReel),
		cmpInt(rightStable, leftStable),
		cmpInt(leftCut, rightCut),
		comparePlanCost(left, right),
	} {
		if comparison != 0 {
			return comparison
		}
	}
	return strings.Compare(left.OrderPlan, right.OrderPlan)
}

func packagingQuantities(plan PurchasePlan) (reel, stable, cut int) {
	for _, leg := range plan.Legs {
		switch packagingKind(leg) {
		case "reel":
			reel += leg.PurchasedQuantity
			stable += leg.PurchasedQuantity
		case "stable":
			stable += leg.PurchasedQuantity
		case "cut":
			cut += leg.PurchasedQuantity
		}
	}
	return reel, stable, cut
}

func packagingKind(leg PurchaseLeg) string {
	text := strings.ToLower(strings.TrimSpace(leg.PackagingMode + " " + leg.PackageType))
	switch {
	case strings.Contains(text, "cut tape"):
		return "cut"
	case strings.Contains(text, "mousereel"), strings.Contains(text, "mouse reel"):
		return "stable"
	case strings.Contains(text, "reel"), strings.Contains(text, "t&r"):
		return "reel"
	case strings.Contains(text, "tray"), strings.Contains(text, "tube"), strings.Contains(text, "bulk"):
		return "stable"
	default:
		return "unknown"
	}
}

func candidateBulkQuantities(
	required int,
	bulk PurchaseFamily,
	remainder PurchaseFamily,
) []int {
	mix := firstPositive(bulk.MixQuantity, bulk.FullReelQuantity, bulk.OrderMultiple)
	if mix <= 1 {
		return nil
	}
	counts := map[int]struct{}{}
	addAround := func(count int) {
		for _, candidate := range []int{count - 1, count, count + 1} {
			if candidate > 0 {
				counts[candidate] = struct{}{}
			}
		}
	}
	addAround(required / mix)
	for _, priceBreak := range bulk.PriceBreaks {
		addAround(priceBreak.Quantity / mix)
	}
	for _, priceBreak := range remainder.PriceBreaks {
		if required > priceBreak.Quantity {
			addAround((required - priceBreak.Quantity) / mix)
		}
	}
	if bulk.AvailableQuantity != nil {
		addAround(*bulk.AvailableQuantity / mix)
	}
	quantities := make([]int, 0, len(counts))
	for count := range counts {
		if count > math.MaxInt/mix {
			continue
		}
		quantity := count * mix
		if quantity > 0 && quantity < required {
			quantities = append(quantities, quantity)
		}
	}
	sort.Ints(quantities)
	return quantities
}

func familyStrategy(family PurchaseFamily, required, purchased, priceBreak int) string {
	switch family.StrategyMode {
	case "static":
		return family.BasePricingStrategy
	case "full_reel":
		return "full reel"
	case "price_break":
		if purchased > required {
			if priceBreak > required {
				return "next price break"
			}
			if family.OrderMultiple > 1 {
				return "order multiple"
			}
			return "next price break"
		}
		if family.BasePricingStrategy != "" {
			return family.BasePricingStrategy
		}
		return "requested quantity"
	default:
		return family.BasePricingStrategy
	}
}

func formatOrderPlan(legs []PurchaseLeg) string {
	rendered := make([]string, 0, len(legs))
	for _, leg := range legs {
		if leg.OrderBatchSize > 1 && leg.OrderBatchCount > 0 {
			noun := "batches"
			if strings.Contains(strings.ToLower(leg.PackagingMode), "reel") {
				noun = "reels"
			}
			if leg.OrderBatchCount == 1 {
				noun = strings.TrimSuffix(noun, "s")
			}
			rendered = append(rendered, fmt.Sprintf(
				"%d %s x %d",
				leg.OrderBatchCount,
				noun,
				leg.OrderBatchSize,
			))
			continue
		}
		if leg.PackagingMode != "" {
			rendered = append(
				rendered,
				fmt.Sprintf("%d %s", leg.PurchasedQuantity, strings.ToLower(leg.PackagingMode)),
			)
			continue
		}
		rendered = append(rendered, fmt.Sprintf("%d", leg.PurchasedQuantity))
	}
	return strings.Join(rendered, " + ")
}

func roundUpToMultiple(quantity, multiple int) (int, error) {
	if multiple <= 1 {
		return quantity, nil
	}
	remainder := quantity % multiple
	if remainder == 0 {
		return quantity, nil
	}
	addition := multiple - remainder
	if quantity > math.MaxInt-addition {
		return 0, errors.New("purchase quantity overflow")
	}
	return quantity + addition, nil
}

func firstDivisor(quantity int, candidates ...int) int {
	for _, candidate := range candidates {
		if candidate > 1 && quantity%candidate == 0 {
			return candidate
		}
	}
	return 0
}

func firstPositive(values ...int) int {
	for _, value := range values {
		if value > 0 {
			return value
		}
	}
	return 0
}

func validCurrency(currency string) bool {
	if len(currency) != 3 {
		return false
	}
	for _, character := range currency {
		if character < 'A' || character > 'Z' {
			return false
		}
	}
	return true
}

func withinBasisPoints(candidate, cheapest money.Decimal, basisPoints int) bool {
	left := new(big.Int).Mul(big.NewInt(candidate.Micros()), big.NewInt(10_000))
	right := new(big.Int).Mul(
		big.NewInt(cheapest.Micros()),
		big.NewInt(int64(10_000+basisPoints)),
	)
	return left.Cmp(right) <= 0
}

func cmpDecimal(left, right money.Decimal) int {
	return cmpInt64(left.Micros(), right.Micros())
}

func cmpInt(left, right int) int {
	switch {
	case left < right:
		return -1
	case left > right:
		return 1
	default:
		return 0
	}
}

func cmpInt64(left, right int64) int {
	switch {
	case left < right:
		return -1
	case left > right:
		return 1
	default:
		return 0
	}
}
